runtime/parachains/src/paras/mod.rs

```diff
@@ -1994,19 +1951,78 @@ impl<T: Config> Pallet<T> {
 		Heads::<T>::insert(para_id, head_data);
 	}
 
-	#[cfg(feature = "runtime-benchmarks")]
-	pub(crate) fn initialize_para_now(id: ParaId, genesis: ParaGenesisArgs) -> DispatchResult {
-		// first queue this para actions..
-		let _ = Self::schedule_para_initialize(id, genesis)?;
+	/// A low-level function to eagerly initialize a given para.
+	pub(crate) fn initialize_para_now(
+		parachains: &mut ParachainsCache<T>,
+		id: ParaId,
+		genesis_data: &ParaGenesisArgs,
+	) {
+		if genesis_data.parachain {
+			parachains.add(id);
+			ParaLifecycles::<T>::insert(&id, ParaLifecycle::Parachain);
+		} else {
+			ParaLifecycles::<T>::insert(&id, ParaLifecycle::Parathread);
+		}
+
+		// HACK: see the notice in `schedule_para_initialize`.
+		//
+		// Apparently, this is left over from a prior version of the runtime.
+		// To handle this we just insert the code and link the current code hash
+		// to it.
+		if !genesis_data.validation_code.0.is_empty() {
+			let code_hash = genesis_data.validation_code.hash();
+			Self::increase_code_ref(&code_hash, &genesis_data.validation_code);
+			CurrentCodeHash::<T>::insert(&id, code_hash);
+		}
+
+		Heads::<T>::insert(&id, &genesis_data.genesis_head);
+	}
+}
+
+/// An overlay over the `Parachains` storage entry that provides a convenient interface for adding
+/// or removing parachains in bulk.
+pub(crate) struct ParachainsCache<T: Config> {
+	// `None` here means the parachains list has not been accessed yet, nevermind modified.
+	parachains: Option<Vec<ParaId>>,
+	_config: PhantomData<T>,
+}
 
-		// .. and immediately apply them.
-		Self::apply_actions_queue(Self::scheduled_session());
+impl<T: Config> ParachainsCache<T> {
+	pub fn new() -> Self {
+		Self { parachains: None, _config: PhantomData }
+	}
 
-		// ensure it has become a para.
-		ensure!(
-			ParaLifecycles::<T>::get(id) == Some(ParaLifecycle::Parachain),
-			"Parachain not created properly"
-		);
-		Ok(())
+	fn ensure_loaded(&mut self) -> &mut Vec<ParaId> {
+		if self.parachains.is_none() {
+			self.parachains = Some(Parachains::<T>::get());
+		}
+		self.parachains
+			.as_mut()
+			.expect("ensured above that the parachains list is Some; qed")
```

Comment:
https://doc.rust-lang.org/std/option/enum.Option.html#method.get_or_insert might come in handy